scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,669 @@
+        """Wrapper to the objective function
+        """
+        self._x = np.array(x)
+        self._energy(self._x)
+        return self._etot
+
+    def _smooth_search(self):
+        """Local search implementation using setulb core function
+        """
+        m = 5
+        iteration = 0
+        f = 0
+        n = self._x.size
+        ndb = np.zeros(self._x.size)
+        ndb.fill(2)
+        iprint = -1
+        x = np.array(self._xbuffer, np.float64)
+        f = np.array(0.0, np.float64)
+        l = np.array(self._lower, np.float64)
+        u = np.array(self._upper, np.float64)
+        g = np.zeros(n, np.float64)
+        wa = np.zeros(2 * m * n + 5 * n + 11 * m * m + 8 * m, np.float64)
+        iwa = np.zeros(3 * n, np.int32)
+        task = np.zeros(1, 'S60')
+        csave = np.zeros(1, 'S60')
+        lsave = np.zeros(4, np.int32)
+        isave = np.zeros(44, np.int32)
+        dsave = np.zeros(29, np.float64)
+        task[:] = 'START'
+        if self.itsoftmax < 100:
+            self.itsoftmax = 100
+        elif self.itsoftmax > 1000:
+            self.itsoftmax = 1000
+        while True:
+            if iteration >= self.itsoftmax:
+                self._xbuffer = np.array(x)
+                self._fvalue = f
+                return
+            _lbfgsb.setulb(m, x, l, u, ndb,
+                           f, g, self.factr, self.pgtol, wa, iwa, task,
+                           iprint, csave, lsave, isave, dsave, 100)
+            iteration += 1
+            task_str = task.tostring()
+            if task_str.startswith(b'FG'):
+                self._xbuffer = np.array(x)
+                f = self._fobjective(self._xbuffer)
+                if self.know_real:
+                    if f <= self.real_threshold:
+                        self._xbuffer = np.array(x)
+                        self._fvalue = f
+                        return
+                g = np.array(self._compute_gradient(x), np.float64)
+            elif task_str.startswith(b'NEW_X'):
+                pass
+            else:
+                self._fvalue = f
+                self._xbuffer = np.array(x)
+                return
+
+    def _compute_gradient(self, x):
+        """Computation of numerical derivatives for local search
+        """
+        g = np.zeros(self._x.size, np.float64)
+        for i in range(self._x.size):
+            x1 = np.array(x)
+            x2 = np.array(x)
+            respl = self.reps
+            respr = self.reps
+            x1[i] = x[i] + respr
+            if x1[i] > self._upper[i]:
+                x1[i] = self._upper[i]
+                respr = x1[i] - x[i]
+            x2[i] = x[i] - respl
+            if x2[i] < self._lower[i]:
+                x2[i] = self._lower[i]
+                respl = x[i] - x2[i]
+            f1 = self._fobjective(x1)
+            f2 = self._fobjective(x2)
+            g[i] = ((f1 - f2)) / (respl + respr)
+        idx = np.logical_or(np.isnan(g), np.isinf(g))
+        g[idx] = 101.0
```

Comment:
Because 42 ;-)
I do not remember, I will check this and let you know... A remaining thing from our old initial FORTRAN code.